core/string/src/builder.rs

```diff
@@ -0,0 +1,713 @@
+use crate::{
+    alloc_overflow, tagged::Tagged, JsStr, JsStrVariant, JsString, RawJsString, RefCount,
+    TaggedLen, DATA_OFFSET,
+};
+
+use std::{
+    alloc::{alloc, dealloc, realloc, Layout},
+    cell::Cell,
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::AddAssign,
+    ptr::{self, addr_of_mut, NonNull},
+    str::{self},
+};
+
+#[doc(hidden)]
+mod private {
+    pub trait Sealed {}
+
+    impl Sealed for u8 {}
+    impl Sealed for u16 {}
+}
+
+/// Inner elements represented for `JsStringBuilder`.
+pub trait JsStringData: private::Sealed {}
+
+impl JsStringData for u8 {}
+impl JsStringData for u16 {}
```

Comment:
Doing a second pass I realized that maybe this abstraction is not needed? More so because pushing `u16` is always valid, but pushing `u8` needs to do latin1 checks.

Since we now use an enum for the generic `StringBuilder` that doesn't need the `JsStringData` trait, maybe we can just have different types for each builder, then just extract the common paths into utility functions.

Reply:
> Doing a second pass I realized that maybe this abstraction is not needed?

`JsStringData` could be removed since I only export specific builders.

> maybe we can just have different types for each builder, then just extract the common paths into utility functions.

Have we already done this by using generic?

> More so because pushing u16 is always valid, but pushing u8 needs to do latin1 checks, and having separate builders for each type would allow us to reflect that in the API.

This could be done with implementations for generics seperately.

Reply:
Yeah, I think that could potentially work.